db-client-java/src/main/java/com/eventstore/dbclient/NodeSelector.java

```diff
@@ -0,0 +1,72 @@
+package com.eventstore.dbclient;
+
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Random;
+import java.util.Set;
+
+class NodeSelector {
+
+    private static final Random random = new Random();
+    private static final Set<ClusterInfo.MemberState> invalidStates;
+    private static final Map<NodePreference,Comparator<ClusterInfo.Member>> memberComparators;
+
+    static {
+        invalidStates = new HashSet<ClusterInfo.MemberState>() {{
```

Comment:
**CONCERN:** I think that: `PreReadOnlyReplica` and `ReadOnlyLeaderless` might be valid state for the `ReadOnlyReplica` preference.

Reply:
@oskardudycz 
I agree, that this might be possible but then it is a different issue as this implementation is based on status quo in trunk:
https://github.com/EventStore/EventStoreDB-Client-Java/blob/trunk/db-client-java/src/main/java/com/eventstore/dbclient/EventStoreDBClusterClient.java#L22
https://github.com/EventStore/EventStoreDB-Client-Java/blob/trunk/db-client-java/src/main/java/com/eventstore/dbclient/EventStoreDBClusterClient.java#L94

Reply:
Yup, I think that we could expand it in the follow-up PR together with other alignments I mentioned inhttps://github.com/EventStore/EventStoreDB-Client-Java/pull/101#pullrequestreview-754754366. This is more than good enough for now :)